Add additional contextual information to invalid-style-prop warning

niole opened the pull request:
inline style prop error thrown for JSX React elements now names React class that erroring element is a child of.  Fixes #4168

niole commented:
This works in a live setting when inside of a working app.

jimfb commented on src/renderers/dom/shared/ReactDOMComponent.js:
```diff
@@ -263,7 +263,7 @@ function assertValidProps(component, props) {
   invariant(
     props.style == null || typeof props.style === 'object',
-    'The `style` prop expects a mapping from style properties to values, ' +
+    'The `style` prop of '+component._currentElement._owner._instance.__proto__.constructor.displayName + ' expects a mapping from style properties to values, ' +
```
It's the style prop of the owner component?  This doesn't look right to me.

niole replied:
I read the issue as being it would be helpful if the error message
contained the name of the react class that has the render method where the
style prop has bad syntax, which is what that piece of code returns
On Jul 24, 2015 11:03 PM, "Jim" <notifications@github.com> wrote:

> In src/renderers/dom/shared/ReactDOMComponent.js
> <https://github.com/facebook/react/pull/4482#discussion_r35480276>:
>
> > @@ -263,7 +263,7 @@ function assertValidProps(component, props) {
> >    }
> >    invariant(
> >      props.style == null || typeof props.style === 'object',
> > -    'The `style` prop expects a mapping from style properties to values, ' +
> > +    'The `style` prop of '+component._currentElement._owner._instance.__proto__.constructor.displayName + ' expects a mapping from style properties to values, ' +
>
> It's the style prop of the owner component? This doesn't look right to me.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/facebook/react/pull/4482/files#r35480276>.
>

niole replied:
Woops must have been sleepy when I submitted that. I would change it to say
'check the render method of '+code_snippet

I read the issue as being it would be helpful if the error message
contained the name of the react class that has the render method where the
style prop has bad syntax, which is what that piece of code returns
On Jul 24, 2015 11:03 PM, "Jim" <notifications@github.com> wrote:

> In src/renderers/dom/shared/ReactDOMComponent.js
> <https://github.com/facebook/react/pull/4482#discussion_r35480276>:
>
> > @@ -263,7 +263,7 @@ function assertValidProps(component, props) {
> >    }
> >    invariant(
> >      props.style == null || typeof props.style === 'object',
> > -    'The `style` prop expects a mapping from style properties to values, ' +
> > +    'The `style` prop of '+component._currentElement._owner._instance.__proto__.constructor.displayName + ' expects a mapping from style properties to values, ' +
>
> It's the style prop of the owner component? This doesn't look right to me.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/facebook/react/pull/4482/files#r35480276>.
>

jimfb commented on src/renderers/dom/shared/ReactDOMComponent.js:
```diff
@@ -263,7 +263,9 @@ function assertValidProps(component, props) {
     props.style == null || typeof props.style === 'object',
-    'The `style` prop expects a mapping from style properties to values, ' +
+    'Please check the render method of ' +
+     component._currentElement._owner._instance.__proto__.constructor.displayName +
```
You need to have an if-check, because there is no guarantee that _owner isn't null (element could have been created outside of a render method).

zpao commented on src/renderers/dom/shared/ReactDOMComponent.js:
```diff
@@ -263,7 +263,9 @@ function assertValidProps(component, props) {
+     component._currentElement._owner._instance.__proto__.constructor.displayName +
+    '. The `style` prop expects a mapping from style properties to values, ' +
     'not a string. For example, style={{marginRight: spacing + \'em\'}} when ' +
     'using JSX.'
```
Can you just use `getDeclarationErrorAddendum` here. That will wrap up the `_owner` stuff and get the name.

zpao commented on src/renderers/dom/shared/ReactDOMComponent.js:
```diff
@@ -263,7 +263,8 @@ function assertValidProps(component, props) {
+      getDeclarationErrorAddendum(component) +
+    '. The `style` prop expects a mapping from style properties to values, ' +
     'not a string. For example, style={{marginRight: spacing + \'em\'}} when ' +
     'using JSX.'
```
Add `%s` into the string here, then pass `getDecl..()` as a 3rd argument.

niole commented:
@jimfb do you think this test is failing bc of the characters used for string interpolation in the non-test version of the error message?

jimfb commented:
@niole I only glanced at it briefly, but I think your test error message may be missing the prefix 'Invariant Violation:' appended to invariants before throwing.  It's often helpful to copy-paste the message into a monospaced text editor, line the messages up, and see where they differ.

zpao commented on src/renderers/dom/shared/ReactDOMComponent.js:
```diff
@@ -263,10 +263,11 @@ function assertValidProps(component, props) {
     'not a string. For example, style={{marginRight: spacing + \'em\'}} when ' +
-    'using JSX.'
-  );
+    'using JSX.',
```
`%s` should be at the end of the string.

zpao commented on src/renderers/dom/shared/ReactDOMComponent.js:
```diff
@@ -263,10 +263,11 @@ function assertValidProps(component, props) {
-    'using JSX.'
-  );
+    'using JSX.',
+     getDeclarationErrorAddendum(component));
```
Please leave the closing paren on its own line.

zpao commented:
And yes, you're missing the 'Invariant Violation:' prefix.

niole commented:
Cool. I'm on it.
On Jul 31, 2015 10:36 AM, "Paul O’Shannessy" <notifications@github.com>
wrote:

> And yes, you're missing the 'Invariant Violation:' prefix.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/facebook/react/pull/4482#issuecomment-126763855>.
>

niole commented:
Should I add this prefix to the test and the actual code?
On Jul 31, 2015 10:41 AM, "Niole Nelson" <niolenelson@gmail.com> wrote:

> Cool. I'm on it.
> On Jul 31, 2015 10:36 AM, "Paul O’Shannessy" <notifications@github.com>
> wrote:
>
>> And yes, you're missing the 'Invariant Violation:' prefix.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/facebook/react/pull/4482#issuecomment-126763855>.
>>
>

jimfb commented:
@niole No, just the test.  The issue is that the actual code adds it automatically but the test does not.

zpao commented on src/renderers/dom/shared/ReactDOMComponent.js:
```diff
@@ -263,9 +263,11 @@ function assertValidProps(component, props) {
   invariant(
     props.style == null || typeof props.style === 'object',
-    'The `style` prop expects a mapping from style properties to values, ' +
+    ' A component inside its render method has incorrect inline style syntax.' +
```
You still have a leading space inside the text which doesn't belong. But do we really need this new sentence? I don't think it really adds anything and with the addendum we'll say to check the render method of a component.

zpao commented on src/renderers/dom/shared/ReactDOMComponent.js:
```diff
@@ -263,9 +263,11 @@ function assertValidProps(component, props) {
+     ' The `style` prop expects a mapping from style properties to values, ' +
     'not a string. For example, style={{marginRight: spacing + \'em\'}} when ' +
-    'using JSX.'
+    'using JSX. %s',
```
Don't put a space before `%s`. If you look at what `getDeclarationErrorAddendum` is doing, it prefixes with a space.

zpao commented:
Please run your tests locally before pushing. You can iterate much more quickly that way instead of waiting for travis to tell you there are failures.

jimfb commented:
You can run your tests using `npm test`.

If you read somewhere that you should use `npm run jest`, be aware that has been removed in favor of `npm test`.

niole commented:
What's up with this eslint-plugin-react-internal and how do I get it?
On Jul 31, 2015 10:58 AM, "Jim" <notifications@github.com> wrote:

> You can run your tests using npm test.
>
> If you read somewhere that you should use npm run jest, be aware that has
> been removed in favor of npm test.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/facebook/react/pull/4482#issuecomment-126771423>.
>

zpao commented:
`npm install` should do it. You may need to install npm >= 2.0 though for that to work. If you're on npm < 2 you might be able to do `npm install ./eslint-rules` to pick it up.

niole commented:
@zpao @jimfb how does it look.

zpao commented on src/renderers/dom/shared/ReactDOMComponent.js:
```diff
@@ -263,9 +263,11 @@ function assertValidProps(component, props) {
   invariant(
     props.style == null || typeof props.style === 'object',
-    'The `style` prop expects a mapping from style properties to values, ' +
+    'A dom node has incorrect inline style syntax.' +
```
Please just leave the beginning of this message as it was. We really just need the addendum.

zpao commented on src/renderers/dom/shared/ReactDOMComponent.js:
```diff
@@ -263,9 +263,11 @@ function assertValidProps(component, props) {
+     ' The `style` prop expects a mapping from style properties to values, ' +
     'not a string. For example, style={{marginRight: spacing + \'em\'}} when ' +
-    'using JSX.'
+    'using JSX. %s',
```
The addendum will start with a space, you should not be adding one yourself here.

zpao commented on src/renderers/dom/shared/__tests__/ReactDOMComponent-test.js:
```diff
@@ -718,9 +718,9 @@ describe('ReactDOMComponent', function() {
-        'style={{marginRight: spacing + \'em\'}} when using JSX.'
+        'Invariant Violation: A dom node has incorrect inline style syntax.' +
+        ' The `style` prop expects a mapping from style properties to values, not' +
+        ' a string. For example, style={{marginRight: spacing + \'em\'}} when using JSX. '
```
Please add a new test confirming that the addendum is actually added when you expect it to be.

zpao commented:
Sorry for the delay, GitHub doesn't send notifications when there are new / changed commits added.

niole commented:
@zpao @jimfb how does it look? I'm also wondering about how to run tests locally. I upgraded npm and everything's installing.
* I then run `npm test` and get `Cannot find module 'fbjs/scripts/babel/rewrite-requires'`. 
 babel, which 'couldn't be found', and a lot of failed tests
* I then upgraded jest to 5.x.
* Now I have this  `jsdom 4.x onward only works on io.js, not Node.js™`, and what seems like less failed tests

should I use io.js?

niole commented:
@zpao @jimfb Does this need work?

zpao commented:
I squashed, cleaned up the unrelated changes, and pushed (your commit: 70c07c67b74b7c62c524f80f7e7b27a8e09eb7f1, merge: bdc7ce9f01bce3b4ebf363fcef20dd3534e614ca). Thanks for sticking with it!

niole commented:
Thanks, Paul!

On Thu, Aug 27, 2015 at 6:29 PM, Paul O’Shannessy <notifications@github.com>
wrote:

> I squashed, cleaned up the unrelated changes, and pushed (your commit:
> 70c07c6
> <https://github.com/facebook/react/commit/70c07c67b74b7c62c524f80f7e7b27a8e09eb7f1>,
> merge: bdc7ce9
> <https://github.com/facebook/react/commit/bdc7ce9f01bce3b4ebf363fcef20dd3534e614ca>).
> Thanks for sticking with it!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/facebook/react/pull/4482#issuecomment-135599033>.
>
